mitodl/micromasters
# recreate_index breakdown in celery subtasks

#### Pre-Flight checklist

⚠️  **NOTE:** This PR is rebased on https://github.com/mitodl/micromasters/pull/4960


#### What are the relevant tickets?
https://github.com/mitodl/micromasters/issues/4959

#### What's this PR do?
✔️  **NOTE:** The implementation is inspired by [reindexing](https://github.com/mitodl/open-discussions/blob/master/search/tasks.py#L699) in Open-Discussions

- Converts the `recreate_index` into a celery task
- Divides the indexing operation into possible celery subtasks to run parallel
- Introduces `ELASTICSEARCH_INDEXING_CHUNK_SIZE` and runs multiple celery tasks of  Program User Enrollment and Percolate Queries
- Each subtask indices `ELASTICSEARCH_INDEXING_CHUNK_SIZE` this many elements


#### How should this be manually tested?
- Run the `recreate_index` command, this should now start a celery task that further starts 2 types of celery subtasks for indexing enrollments and percolates. 
- Set a test value for `ELASTICSEARCH_INDEXING_CHUNK_SIZE` (Default is 100)
- On running the command you should see multiple celery tasks sent to perform indexing on chunks of data(chunk size would be `ELASTICSEARCH_INDEXING_CHUNK_SIZE`)
- Make sure that the number of celery tasks matches the chunk size division above
- Check the parent task logs the task details which are traceable. e.g if the task fails we can investigate


#### Where should the reviewer start?
- `recreate_index` command

#### Any background context you want to provide?
- We used to run reindexing as a synchronous process before, and we were facing timeouts during the process when the recreate_index was run on a lot of data
- We did some related updates in [Open Discussions](https://github.com/mitodl/open-discussions/blob/master/search/tasks.py#L699) and it made sense to update micromasters also to make reindexing faster
